src/ruby/puppetserver-lib/puppet/server/ast_compiler.rb

```diff
@@ -24,26 +24,98 @@ def self.compile(compile_options)
+          require 'bolt'
+          require 'bolt/target'
+          require 'bolt/apply_inventory'
+          require 'bolt/apply_target'
+          require 'bolt/apply_result'
+          require 'addressable'
+
+          Puppet[:rich_data] = true
+          Puppet[:node_name_value] = compile_options['certname']
+
+          boltlib_path = compile_options.dig('options', 'boltlib_path')
+          # Use the existing environment with the requested name
+          Puppet::Pal.in_environment(compile_options['environment'],
+                                     pre_modulepath: boltlib_path,
+                                     envpath: Puppet[:environmentpath],
+                                     facts: compile_options['facts']['values']) do |pal|
+            # TODO: We need to decide on an appropriate config for inventory. Given we 
+            # hide this from plan authors this current iteration has only the "required"
+            # data for now. This is being discussed on https://github.com/puppetlabs/bolt/pull/1770
+            fake_config = {
+              'transport' => 'redacted',
+              'transports' => {
+                'redacted' => 'redacted'
+              }
+            }
+            bolt_inv = Bolt::ApplyInventory.new(fake_config)
+            Puppet.override(bolt_inventory: bolt_inv) do
+              Puppet.lookup(:pal_current_node).trusted_data = compile_options['trusted_facts']['values']
+              # This compiler has been configured with a node containing
+              # the requested environment, facts, and variables, and is used
+              # to compile a catalog in that context from the supplied AST.
+              pal.with_catalog_compiler do |compiler|
+                #TODO: Should we update these in PAL? They are the defaults in Puppet
+                Puppet[:strict] = :warning
+                Puppet[:strict_variables] = false
+                vars = Puppet::Pops::Serialization::FromDataConverter.convert(compile_options['variables']['values'])
+
+                compile_options['facts']['values'].keys.each {|fact_name| vars.delete(fact_name)}
+                # TODO: Refactor PAL api such that we do not need to use private methods
+                pal.send(:add_variables, compiler.send(:topscope), vars)
+
+                # We have to parse the AST inside the compiler block, because it
+                # initializes the necessary type loaders for us.
+                ast = Puppet::Pops::Serialization::FromDataConverter.convert(code)
+                Puppet.warning("AST IS A: Puppet::Pops::Model::Program")
+                # Node definitions must be at the top level of the apply block.
+                # That means the apply body either a) consists of just a
+                # NodeDefinition, b) consists of a BlockExpression which may
+                # contain NodeDefinitions, or c) doesn't contain NodeDefinitions.
+                definitions = if ast.is_a?(Puppet::Pops::Model::BlockExpression)
```

Comment:
It might be valuable, besides adding PUP tickets to the bits that need to be upstreamed into Puppet, adding links to the tickets/issues/code in Bolt that these are copying.